README.md
[![GitHub license](https://dmlc.github.io/img/apache2.svg)](./LICENSE)
[![CircleCI](https://circleci.com/gh/cornell-zhang/heterocl/tree/master.svg?style=svg&circle-token=2b5ee9faf30b94aac41b61032d03e4654a65079d)](https://circleci.com/gh/cornell-zhang/heterocl/tree/master)

HeteroCL: A Multi-Paradigm Programming Infrastructure for Software-Defined Reconfigurable Computing
===================================================================================================

[Website](http://heterocl.csl.cornell.edu/web/index.html) | [Installation](http://heterocl.csl.cornell.edu/doc/installation.html) | [Tutorials](http://heterocl.csl.cornell.edu/doc/tutorials/index.html) | [Samples](http://heterocl.csl.cornell.edu/doc/samples/index.html) | [Documentation](http://heterocl.csl.cornell.edu/doc/index.html)

## Introduction

With the pursuit of improving compute performance under strict power constraints, there is an increasing need for deploying applications to heterogeneous hardware architectures with accelerators, such as GPUs and FPGAs. However, although these heterogeneous computing platforms are becoming widely available, they are very difficult to program especially with FPGAs. As a result, the use of such platforms has been limited to a small subset of programmers with specialized hardware knowledge.

To tackle this challenge, we introduce HeteroCL, a programming infrastructure comprised of a Python-based domain-specific language (DSL) and a compilation flow. The HeteroCL DSL provides a clean programming abstraction that decouples algorithm specification from three important types of hardware customization in compute, data types, and memory architectures. HeteroCL can further capture the interdependence among these different customization techniques, allowing programmers to explore various performance/area/accuracy trade-offs in a systematic and productive manner. In addition, our framework currently provides two advanced domain-specific optimizations with stencil analysis and systolic array generation, which produce highly efficient microarchitectures for accelerating popular workloads from image processing and deep learning domains.

## Current Compilation Flow

![flow](docs/github-flow.png)

## Evaluation on AWS F1 (Xilinx Virtex UltraScale+<sup>TM</sup> VU9P FPGA)
The speedup is over a single-core single-thread CPU execution on AWS F1.

| Benchmark | Data Sizes & Type | #LUTs | #FFs | #BRAMs | #DSPs | Freqency (MHz) | Speedup | Back End |
| :-------- | :---------------- | :----: | :----:| :-----: | :----: | :------------: | :------:| :------: |
| **KNN Digit Recognition**<br/>Image classification | K=3 #images=1800<br/>`uint49` | 4009 | 5835 | 88 | 0 | 250 | 12.5 | General |
| **K-Means**<br/> Clustering | K=16 #elem=320 x 32<br/>`int32` | 212708 | 235011 | 32 | 1536 | 190.6 | 16.0 | General |
| **Smith-Waterman**<br/> Genomic sequencing | string len=128<br/>`uint2` | 110841 | 88369 | 1409 | 0 | 152.2 | 20.9 | General |
| **Seidel**<br/> Image processing | 2160 pixel x 3840 pixel<br/>`fixed16` | 21719 | 31663 | 46 | 96 | 250 | 5.9 | Stencil |
| **Gaussian**<br/> Image processing | 2160 pixel x 3840 pixel<br/>`fixed16` | 70833 | 131160 | 46 | 688 | 250 | 13.2 | Stencil |
| **Jacobi**<br/> Linear algebra | 2160 pixel x 3840 pixel<br/>`fixed16` | 14883 | 22485 | 46 | 48 | 250 | 5.0 | Stencil |
| **GEMM**<br/> Matrix multiplication | 1024 x 1024 x 1024<br/>`fixed16` | 454492 | 800283 | 932 | 2507 | 236.8 | 8.9 | Systolic Array |
| **LeNet Inference**<br/> CNN | MNIST<br/>`fixed16` | 362291 | 660186 | 739.5 | 1368 | 250 | 10.6 | Systolic Array |


## Publication

If you use HeteroCL in your design, please cite our [FPGA'19 paper](http://www.csl.cornell.edu/~zhiruz/pdfs/heterocl-fpga2019.pdf):
```
@article{lai2019heterocl,
  title={HeteroCL: A Multi-Paradigm Programming Infrastructure for Software-Defined Reconfigurable Computing},
  author={Lai, Yi-Hsiang and Chi, Yuze and Hu, Yuwei and Wang, Jie and Yu, Cody Hao and 
          Zhou, Yuan and Cong, Jason and Zhang, Zhiru},
  journal={Int'l Symp. on Field-Programmable Gate Arrays (FPGA)},
  year={2019}
}
```

## Related Work

HeteroCL is a Python-based DSL extended from TVM and it extends Halide IR for intermediate representation. HeterCL incoporates the SODA framework, PolySA framework, and Merlin Compiler for FPGA back-end generation.

* **[Stencil with Optimized Dataflow Architecture](https://vast.cs.ucla.edu/~chiyuze/pub/iccad18.pdf)** (SODA)
* **[Polyhedral-Based Systolic Array Auto-Compilation](http://cadlab.cs.ucla.edu/~jaywang/papers/iccad18-polysa.pdf)** (PolySA)
* **[Merlin Compiler](https://www.falconcomputing.com/merlin-fpga-compiler/)**
* **[Halide](https://halide-lang.org)**
* **[TVM](https://tvm.ai)**

## Contributing to HeteroCL
1. Use [Pull Request](https://help.github.com/articles/about-pull-requests/).
2. Python [coding style](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles).
3. Python [docstring style](https://numpydoc.readthedocs.io/en/latest/format.html#other-points-to-keep-in-mind).


HISTORY
### TODO
  * consider data movement in nested loops. 
  * consider "move data from local to kernel"? 
  * correct location to insert "attr partition" in SDAccel 
  * move data to different storage on xcel?
  * generate the host code & Makefiel for simulation

    * 1. memory object allocation + copy data + enqueue tasks before top_function and data fetching at the end of the function call (Important: To handle pipes between kernels, different kernels but sub functions are required)

    * 2. do we still want to reuse opencl utility functions from rosetta? it's not portable and fine-granular enough. And also it requries extra efforst to pass the IR information (e.g. size, data type) to the wrapper code generater. The solution is to push OpenCL host code to a separate code generator inherited from CodeGenC. and create initilization code for Kernel Stmt IR.

    * 3. update the simulation flow : we separate the simultaion to two stages a) generate wrapper code and compile. The wrapper code copies the data from python front-end to the execution engine (e.g. TVM PackedFunc) b) update and execute. The array args are automaticallly copied using the shared memory initialized in wrapper code. However, the scalar varible needs to be update. **To summarize**, the simulation enables a one-time compilation, which frees the users from compiling the program everytime when calling the function.

# 2020-01-31
  * create reusable conv2d function in hlib
  * fixed clang seg fault (kernel name mis-matching)

# 2020-01-30
  * fixed issues of wrong input buffer in schedule_dataflow
  * update the SDAccel codegen
    * 1. macro for opencl 1.2 legacy support
    * 2. remove `__local` identifier for allocate ir 
    * 3. check the sequence of kernel def string 
    * 4. for variable type arg in top func ()

# 2020-01-28
  * udpate the HLSC & OpenCL codegen for SDAccel  
    * 1. channel read & write on host and top_function on device (ignore the nested for loops to transmit data through chanels) 
    * 2. channels between kernel functions (create opencl channels in dataflow mode where one kernel function, which is decorated with dataflow attribute, wrapping multiple sub-functions + blocking pipes for multiple kernels)
    * 3. KernelDef IR printer for SDAccel (__kernel attribute + dataflow)

# 2020-01-25
  * Updated streaming ir pass to perform device scope grouping & the compute being offloaded is modelled as a KernelDef and corresponding KernelStmt (How to manage the streaming channel allocation between host & device: create two identical buffer in host & device)

# 2020-01-21
  * Handle Host Device Splitting (How to generate the top function calls from the host. Refer to HostDeviceSplitter in TVM to mutate all body in attr stmts marked with thread_extent or pipeline_exec_scope. These stmts are considered as Call IR nodes and wrapped into LoweredFunc) 
    * 1. modify the stream inference ir pass (first analyze all the device_scop attr and use a unified attr stmt to group all adjacent stmts. And then split the program with device scope attr stmt body. wrap the body with a Call node with all argument information included)  
    * 2. modified the `move_to` scheduling in HeteroCL (alloc buffer does not match with var. We need to bind the arg to the tensor in StorageFaltten pass)

soda_A_C.cpp
#include "soda_stencil.h"
#include<float.h>
#include<math.h>
#include<stdbool.h>
#include<stddef.h>
#include<stdint.h>
#include<stdio.h>
#include<string.h>
#include<ap_int.h>
#include<hls_stream.h>


#ifndef BURST_WIDTH
#define BURST_WIDTH 512
#endif//BURST_WIDTH

#ifdef UNROLL_FACTOR
#if UNROLL_FACTOR != 1
#error UNROLL_FACTOR != 1
#endif//UNROLL_FACTOR != 1
#endif//UNROLL_FACTOR
#ifdef TILE_SIZE_DIM_0
#if TILE_SIZE_DIM_0 != 8
#error TILE_SIZE_DIM_0 != 8
#endif//TILE_SIZE_DIM_0 != 8
#endif//TILE_SIZE_DIM_0
#ifdef BURST_WIDTH
#if BURST_WIDTH != 512
#error BURST_WIDTH != 512
#endif//BURST_WIDTH != 512
#endif//BURST_WIDTH

template<typename T> struct Data
{
  T data;
  bool ctrl;
};
template<typename To, typename From>
inline To Reinterpret(const From& val)
{
  return reinterpret_cast<const To&>(val);
}
template<typename T> inline bool ReadData(T* data, hls::stream<Data<T>>* from)
{
#pragma HLS inline
  const Data<T>& tmp = from->read();
  *data = tmp.data;
  return tmp.ctrl;
}
template<typename T> inline void WriteData(hls::stream<Data<T>>* to, const T& data, bool ctrl)
{
#pragma HLS inline
  Data<T> tmp;
  tmp.data = data;
  tmp.ctrl = ctrl;
  to->write(tmp);
}
void BurstRead(hls::stream<Data<ap_uint<BURST_WIDTH>>>* to, ap_uint<BURST_WIDTH>* from, uint64_t data_num)
{
load_epoch:
  for (uint64_t epoch = 0; epoch < data_num;)
  {
#pragma HLS pipeline II=1
    const uint64_t next_epoch = epoch + 1;
    WriteData(to, from[epoch], next_epoch < data_num);
    epoch = next_epoch;
  }
}
void BurstWrite(ap_uint<BURST_WIDTH>* to, hls::stream<Data<ap_uint<BURST_WIDTH>>>* from, uint64_t data_num)
{
store_epoch:
  for (uint64_t epoch = 0; epoch < data_num; ++epoch)
  {
#pragma HLS pipeline II=1
    ap_uint<BURST_WIDTH> buf;
    ReadData(&buf, from);
    to[epoch] = buf;
  }
}
void Module0Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /* input*/ hls::stream<Data<ap_uint<512>>>* dram_A_bank_0_fifo)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = dram_A_bank_0_fifo
module_0_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=16
    if (!dram_A_bank_0_fifo->empty())
    {
      ap_uint<512> dram_A_bank_0_buf;
      const bool dram_A_bank_0_buf_enable = ReadData(&dram_A_bank_0_buf, dram_A_bank_0_fifo);
      const bool enabled = dram_A_bank_0_buf_enable;
      enable = enabled;
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(31, 0))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(63, 32))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(95, 64))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(127, 96))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(159, 128))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(191, 160))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(223, 192))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(255, 224))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(287, 256))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(319, 288))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(351, 320))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(383, 352))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(415, 384))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(447, 416))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(479, 448))), true);
      WriteData(fifo_st_0, Reinterpret<int32_t>(static_cast<ap_uint<32>>(dram_A_bank_0_buf(511, 480))), enabled);
    } // if not empty
  } // for module_0_epoch
} // Module0Func
void Module1Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_1, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_st_1
#pragma HLS data_pack variable = fifo_ld_0
module_1_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      WriteData(fifo_st_0, int32_t(fifo_ref_0), enabled);
      WriteData(fifo_st_1, int32_t(fifo_ref_0), enabled);
    } // if not empty
  } // for module_1_epoch
} // Module1Func
void Module2Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_1, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_st_1
#pragma HLS data_pack variable = fifo_ld_0
  int32_t fifo_ref_0_delayed_9_buf[9];
  ap_uint<4> fifo_ref_0_delayed_9_ptr = 0;
module_2_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
#pragma HLS dependence variable=fifo_ref_0_delayed_9_buf inter false
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      const int32_t fifo_ref_0_delayed_9 = fifo_ref_0_delayed_9_buf[fifo_ref_0_delayed_9_ptr];;
      const int32_t let_0 = fifo_ref_0_delayed_9;
      WriteData(fifo_st_0, int32_t(let_0), enabled);
      WriteData(fifo_st_1, int32_t(let_0), enabled);
      fifo_ref_0_delayed_9_buf[fifo_ref_0_delayed_9_ptr] = fifo_ref_0;
      fifo_ref_0_delayed_9_ptr = fifo_ref_0_delayed_9_ptr < 8 ? ap_uint<4>(fifo_ref_0_delayed_9_ptr+1) : ap_uint<4>(0);
    } // if not empty
  } // for module_2_epoch
} // Module2Func
void Module3Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_ld_0
  int32_t fifo_ref_0_delayed_9_buf[9];
  ap_uint<4> fifo_ref_0_delayed_9_ptr = 0;
module_3_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
#pragma HLS dependence variable=fifo_ref_0_delayed_9_buf inter false
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      const int32_t fifo_ref_0_delayed_9 = fifo_ref_0_delayed_9_buf[fifo_ref_0_delayed_9_ptr];;
      const int32_t let_0 = fifo_ref_0_delayed_9;
      WriteData(fifo_st_0, int32_t(let_0), enabled);
      fifo_ref_0_delayed_9_buf[fifo_ref_0_delayed_9_ptr] = fifo_ref_0;
      fifo_ref_0_delayed_9_ptr = fifo_ref_0_delayed_9_ptr < 8 ? ap_uint<4>(fifo_ref_0_delayed_9_ptr+1) : ap_uint<4>(0);
    } // if not empty
  } // for module_3_epoch
} // Module3Func
void Module4Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_1, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_2)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_ld_0
#pragma HLS data_pack variable = fifo_ld_1
#pragma HLS data_pack variable = fifo_ld_2
module_4_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
    if (!fifo_ld_0->empty() && !fifo_ld_1->empty() && !fifo_ld_2->empty())
    {
      int32_t fifo_ref_0;
      int32_t fifo_ref_1;
      int32_t fifo_ref_2;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool fifo_ref_1_enable = ReadData(&fifo_ref_1, fifo_ld_1);
      const bool fifo_ref_2_enable = ReadData(&fifo_ref_2, fifo_ld_2);
      const bool enabled = fifo_ref_0_enable && fifo_ref_1_enable && fifo_ref_2_enable;
      enable = enabled;
      WriteData(fifo_st_0, int32_t(static_cast<int32_t >(static_cast<ap_int<34> >(static_cast<ap_int<33> >(fifo_ref_0) + static_cast<ap_int<33> >(fifo_ref_1)) + static_cast<ap_int<34> >(fifo_ref_2))), enabled);
    } // if not empty
  } // for module_4_epoch
} // Module4Func
void Module5Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_1, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_st_1
#pragma HLS data_pack variable = fifo_ld_0
  int32_t fifo_ref_0_delayed_1_buf[1];
  ap_uint<1> fifo_ref_0_delayed_1_ptr = 0;
module_5_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
#pragma HLS dependence variable=fifo_ref_0_delayed_1_buf inter false
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      const int32_t fifo_ref_0_delayed_1 = fifo_ref_0_delayed_1_buf[fifo_ref_0_delayed_1_ptr];;
      const int32_t let_0 = fifo_ref_0_delayed_1;
      WriteData(fifo_st_0, int32_t(let_0), enabled);
      WriteData(fifo_st_1, int32_t(let_0), enabled);
      fifo_ref_0_delayed_1_buf[fifo_ref_0_delayed_1_ptr] = fifo_ref_0;
      fifo_ref_0_delayed_1_ptr = fifo_ref_0_delayed_1_ptr < 0 ? ap_uint<1>(fifo_ref_0_delayed_1_ptr+1) : ap_uint<1>(0);
    } // if not empty
  } // for module_5_epoch
} // Module5Func
void Module6Func(
  /*output*/ hls::stream<Data<int32_t>>* fifo_st_0, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = fifo_st_0
#pragma HLS data_pack variable = fifo_ld_0
  int32_t fifo_ref_0_delayed_1_buf[1];
  ap_uint<1> fifo_ref_0_delayed_1_ptr = 0;
module_6_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=1
#pragma HLS dependence variable=fifo_ref_0_delayed_1_buf inter false
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      const int32_t fifo_ref_0_delayed_1 = fifo_ref_0_delayed_1_buf[fifo_ref_0_delayed_1_ptr];;
      const int32_t let_0 = fifo_ref_0_delayed_1;
      WriteData(fifo_st_0, int32_t(let_0), enabled);
      fifo_ref_0_delayed_1_buf[fifo_ref_0_delayed_1_ptr] = fifo_ref_0;
      fifo_ref_0_delayed_1_ptr = fifo_ref_0_delayed_1_ptr < 0 ? ap_uint<1>(fifo_ref_0_delayed_1_ptr+1) : ap_uint<1>(0);
    } // if not empty
  } // for module_6_epoch
} // Module6Func
void Module7Func(
  /*output*/ hls::stream<Data<ap_uint<512>>>* dram_C_bank_0_fifo, 
  /* input*/ hls::stream<Data<int32_t>>* fifo_ld_0)
{
#pragma HLS data_pack variable = dram_C_bank_0_fifo
#pragma HLS data_pack variable = fifo_ld_0
module_7_epoch:
  for (bool enable = true; enable;)
  {
#pragma HLS pipeline II=16
    if (!fifo_ld_0->empty())
    {
      int32_t fifo_ref_0;
      ap_uint<512> dram_C_bank_0_buf;
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(31, 0) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(63, 32) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(95, 64) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(127, 96) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(159, 128) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(191, 160) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(223, 192) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(255, 224) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(287, 256) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(319, 288) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(351, 320) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(383, 352) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(415, 384) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(447, 416) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      ReadData(&fifo_ref_0, fifo_ld_0);
      dram_C_bank_0_buf(479, 448) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      const bool fifo_ref_0_enable = ReadData(&fifo_ref_0, fifo_ld_0);
      const bool enabled = fifo_ref_0_enable;
      enable = enabled;
      dram_C_bank_0_buf(511, 480) = Reinterpret<ap_uint<32>>(fifo_ref_0);
      WriteData(dram_C_bank_0_fifo, dram_C_bank_0_buf, enabled);
    } // if not empty
  } // for module_7_epoch
} // Module7Func
extern "C"
{

void soda_A_C_kernel(
  ap_uint<512>* bank_0_C,
  ap_uint<512>* bank_0_A,
  uint64_t coalesced_data_num)
{
#pragma HLS interface m_axi port=bank_0_C offset=slave depth=65536 bundle=C_bank_0
#pragma HLS interface m_axi port=bank_0_A offset=slave depth=65536 bundle=A_bank_0

#pragma HLS interface s_axilite port=bank_0_C bundle=control
#pragma HLS interface s_axilite port=bank_0_A bundle=control
#pragma HLS interface s_axilite port=coalesced_data_num bundle=control
#pragma HLS interface s_axilite port=return bundle=control

  hls::stream<Data<ap_uint<512>>> bank_0_A_buf("bank_0_A_buf");
#pragma HLS stream variable=bank_0_A_buf depth=32
#pragma HLS data_pack variable=bank_0_A_buf
  hls::stream<Data<ap_uint<512>>> bank_0_C_buf("bank_0_C_buf");
#pragma HLS stream variable=bank_0_C_buf depth=32
#pragma HLS data_pack variable=bank_0_C_buf

  hls::stream<Data<int32_t>> from_super_source_to_A_offset_0("from_super_source_to_A_offset_0");
#pragma HLS stream variable=from_super_source_to_A_offset_0 depth=16
#pragma HLS data_pack variable=from_super_source_to_A_offset_0
  hls::stream<Data<int32_t>> from_A_offset_0_to_A_offset_9("from_A_offset_0_to_A_offset_9");
#pragma HLS stream variable=from_A_offset_0_to_A_offset_9 depth=16
#pragma HLS data_pack variable=from_A_offset_0_to_A_offset_9
  hls::stream<Data<int32_t>> from_A_offset_0_to_B_pe_0("from_A_offset_0_to_B_pe_0");
#pragma HLS stream variable=from_A_offset_0_to_B_pe_0 depth=16
#pragma HLS data_pack variable=from_A_offset_0_to_B_pe_0
  hls::stream<Data<int32_t>> from_A_offset_9_to_A_offset_18("from_A_offset_9_to_A_offset_18");
#pragma HLS stream variable=from_A_offset_9_to_A_offset_18 depth=16
#pragma HLS data_pack variable=from_A_offset_9_to_A_offset_18
  hls::stream<Data<int32_t>> from_A_offset_9_to_B_pe_0("from_A_offset_9_to_B_pe_0");
#pragma HLS stream variable=from_A_offset_9_to_B_pe_0 depth=16
#pragma HLS data_pack variable=from_A_offset_9_to_B_pe_0
  hls::stream<Data<int32_t>> from_A_offset_18_to_B_pe_0("from_A_offset_18_to_B_pe_0");
#pragma HLS stream variable=from_A_offset_18_to_B_pe_0 depth=16
#pragma HLS data_pack variable=from_A_offset_18_to_B_pe_0
  hls::stream<Data<int32_t>> from_B_pe_0_to_B_offset_0("from_B_pe_0_to_B_offset_0");
#pragma HLS stream variable=from_B_pe_0_to_B_offset_0 depth=16
#pragma HLS data_pack variable=from_B_pe_0_to_B_offset_0
  hls::stream<Data<int32_t>> from_B_offset_0_to_B_offset_1("from_B_offset_0_to_B_offset_1");
#pragma HLS stream variable=from_B_offset_0_to_B_offset_1 depth=16
#pragma HLS data_pack variable=from_B_offset_0_to_B_offset_1
  hls::stream<Data<int32_t>> from_B_offset_0_to_C_pe_0("from_B_offset_0_to_C_pe_0");
#pragma HLS stream variable=from_B_offset_0_to_C_pe_0 depth=16
#pragma HLS data_pack variable=from_B_offset_0_to_C_pe_0
  hls::stream<Data<int32_t>> from_B_offset_1_to_B_offset_2("from_B_offset_1_to_B_offset_2");
#pragma HLS stream variable=from_B_offset_1_to_B_offset_2 depth=16
#pragma HLS data_pack variable=from_B_offset_1_to_B_offset_2
  hls::stream<Data<int32_t>> from_B_offset_1_to_C_pe_0("from_B_offset_1_to_C_pe_0");
#pragma HLS stream variable=from_B_offset_1_to_C_pe_0 depth=16
#pragma HLS data_pack variable=from_B_offset_1_to_C_pe_0
  hls::stream<Data<int32_t>> from_B_offset_2_to_C_pe_0("from_B_offset_2_to_C_pe_0");
#pragma HLS stream variable=from_B_offset_2_to_C_pe_0 depth=16
#pragma HLS data_pack variable=from_B_offset_2_to_C_pe_0
  hls::stream<Data<int32_t>> from_C_pe_0_to_super_sink("from_C_pe_0_to_super_sink");
#pragma HLS stream variable=from_C_pe_0_to_super_sink depth=16
#pragma HLS data_pack variable=from_C_pe_0_to_super_sink

#pragma HLS dataflow
  BurstRead(&bank_0_A_buf, bank_0_A, coalesced_data_num);
  Module0Func(
    /*output*/ &from_super_source_to_A_offset_0, 
    /* input*/ &bank_0_A_buf);
  Module1Func(
    /*output*/ &from_A_offset_0_to_A_offset_9, 
    /*output*/ &from_A_offset_0_to_B_pe_0, 
    /* input*/ &from_super_source_to_A_offset_0);
  Module2Func(
    /*output*/ &from_A_offset_9_to_A_offset_18, 
    /*output*/ &from_A_offset_9_to_B_pe_0, 
    /* input*/ &from_A_offset_0_to_A_offset_9);
  Module3Func(
    /*output*/ &from_A_offset_18_to_B_pe_0, 
    /* input*/ &from_A_offset_9_to_A_offset_18);
  Module4Func(
    /*output*/ &from_B_pe_0_to_B_offset_0, 
    /* input*/ &from_A_offset_18_to_B_pe_0, 
    /* input*/ &from_A_offset_9_to_B_pe_0, 
    /* input*/ &from_A_offset_0_to_B_pe_0);
  Module1Func(
    /*output*/ &from_B_offset_0_to_B_offset_1, 
    /*output*/ &from_B_offset_0_to_C_pe_0, 
    /* input*/ &from_B_pe_0_to_B_offset_0);
  Module5Func(
    /*output*/ &from_B_offset_1_to_B_offset_2, 
    /*output*/ &from_B_offset_1_to_C_pe_0, 
    /* input*/ &from_B_offset_0_to_B_offset_1);
  Module6Func(
    /*output*/ &from_B_offset_2_to_C_pe_0, 
    /* input*/ &from_B_offset_1_to_B_offset_2);
  Module4Func(
    /*output*/ &from_C_pe_0_to_super_sink, 
    /* input*/ &from_B_offset_2_to_C_pe_0, 
    /* input*/ &from_B_offset_1_to_C_pe_0, 
    /* input*/ &from_B_offset_0_to_C_pe_0);
  Module7Func(
    /*output*/ &bank_0_C_buf, 
    /* input*/ &from_C_pe_0_to_super_sink);
  BurstWrite(bank_0_C, &bank_0_C_buf, coalesced_data_num);
}

}//extern "C"
